### PR TITLE
Decouple `interpolate` functions from trajectory

### PR DESCRIPTION
**Pull request by isVoid**
## Description
closes https://github.com/rapidsai/cuspatial/issues/644

This PR renames `t` argument of `CubicSpline` to indicate that it can be used as general CubicSpline interpolation function.
Other minor documentation refresh are also included.

## Checklist
- [x] I am familiar with the [Contributing Guidelines](https://github.com/rapidsai/cuspatial/blob/HEAD/CONTRIBUTING.md).
- [x] New or existing tests cover these changes.
- [x] The documentation is up to date with these changes.

### CODE REVIEW

File: python/cuspatial/cuspatial/core/trajectory/generic.py [hunk 1]
```diff
@@ -1,4 +1,4 @@
-# Copyright (c) 2019, NVIDIA CORPORATION.
+# Copyright (c) 2019-2022, NVIDIA CORPORATION.
```
Comment (harrism) [inline python/cuspatial/cuspatial/core/trajectory/generic.py:1]: I don't like naming this "generic.py". Why can't you leave it as `trajectory.py`?

**Review by harrism — CHANGES_REQUESTED** [review]

**Comment by isVoid** [comment]
As discussed in weekly, we maintain `trajectory` module and `interpolate` module. The `interpolate` module is refactored to be generic interpolation functions. Specifically, argument `t` is renamed as `x` for general independent variable.

**Comment by isVoid** [comment]
During rewriting the docs, I was trying to run micro benchmarks to find out about the performance cut-off point and raise warnings for users about a "too-small input". But the result is a bit confusing as the cut off point seems to shift by the ratio of the number of curves and the complexity of the curve. See benchmark: https://gist.github.com/isVoid/108d388173e7199758e31c1ce7f6842a

Perhaps @thomcom can help explain?

**Comment by thomcom** [comment]
Hey @isVoid I'm not sure what you mean by the cut-off point?

**Comment by isVoid** [comment]
@thomcom currently, in `CubicSpline` code we intentionally to avoid API parity with scipy because we want to avoid user calling the API when the input data size is small: https://github.com/rapidsai/cuspatial/blob/19a8d3fada453f411d1b29a6d9449bc1c0e0c2fe/python/cuspatial/cuspatial/core/interpolate.py#L45-L47

This isn't helpful when user still calls the API with a small data size without aware of the performance implication. In this refactor I attempt to argue that the proper way to design the API is to use the same argument variable as scipy and raise a `UserWarning` when input size is small. This is similar to how cudf `groupby.apply` handles many group cases. The key question here is: how to determine when the input size is too small? There must exist some data size when scipy is more performant than cuspatial, and that data size is the "cut-off" point I mentioned above.

**Comment by thomcom** [comment]
Ah, I see. I just ran a few quick benchmarks and I see that scipy is 15x faster than cuspatial on interpolations on my HP Z8 workstation. cuspatial only parallelizes each individual spline fit, so I estimate that if you've got 15 trajectories, cuspatial will be faster at interpolation. I was able to verify this with a set of 15 trajectories of length 1m. Creating the curve object still takes 15x as long (which is basically just creating 15 interpolations in parallel, each taking 15x as long as the hcurve) but interpolating all 15m samples takes 8ms, interpolating 15m samples in scikit learn takes 230ms.

**Comment by isVoid** [comment]
> a set of 15 trajectories of length 1m

Do you mean 15 curves, each curve has 1,000,000 vertices?

**Comment by thomcom** [comment]
> > a set of 15 trajectories of length 1m
> 
> Do you mean 15 curves, each curve has 1,000,000 vertices?

Yes

> Do you mean when cuspatial interpolates 15 curves in parallel, the total time of the kernel is 15x as long as scipy? What's the order of time cost here, `s`? `ms`?

I think that when cuspatial interpolates 15 curves in parallel the cost is similar to using scipy to compute them in serial.

> Sounds like cuspatial has advantage when sampling a complex curve (many-vertices curve) and does not have advantage computing the curve parameters when the total number of curve is small? I think we can raise a performance warning when we see small number of curves (<100) is constructed, do you agree?

I was thinking that < 15 curves was appropriate to compute the warning. I'll write a notebook to concrete the benchmark.

> Out of scope, but can you fit the curve with scipy and pass the parameter to device and sample with device? If you only have 15 curves, you only have 15x4 parameters.

This would be theoretically possible but our curve fitting equations are not identical to scipys, iirc.

**Comment by isVoid** [comment]
rerun tests

File: python/cuspatial/cuspatial/core/interpolate.py [hunk 2]
```diff
@@ -36,87 +36,82 @@ class CubicSpline:
...
+    -------
+    CubicSpline : callable `o`
+        ``o.c`` contains the coefficients that can be used to compute new
+        points along the spline fitting the original ``t`` data. ``o(n)``
+        interpolates the spline coordinates along new input values ``n``.
+
+    Note
+    ----
+    cuSpatial will outperform scipy when many splines are
     fit simultaneously. Data must be arranged in a SoA format, and the
```
Comment (harrism) [inline python/cuspatial/cuspatial/core/interpolate.py:70]: SoA may be unclear to some readers...

**Review by harrism — APPROVED** [review]

File: python/cuspatial/cuspatial/core/interpolate.py (hunk 2 above)
Comment (isVoid) [reply python/cuspatial/cuspatial/core/interpolate.py]: Added full name for readability.

File: python/cuspatial/cuspatial/core/interpolate.py [hunk 3]
```diff
@@ -67,9 +67,9 @@ class CubicSpline:
     Note
     ----
     cuSpatial will outperform scipy when many splines are
-    fit simultaneously. Data must be arranged in a SoA format, and the
-    exclusive `prefix_sum` of the separate curves must also be passed to the
-    function. See example for detail.
+    fit simultaneously. Data must be arranged in a structure of arrays (SoA)
+    format, and the exclusive `prefix_sum` of the separate curves must also be
```
Comment (harrism) [inline python/cuspatial/cuspatial/core/interpolate.py:72]: These are the offsets to the . Don't call it a prefix sum, call it offsets. Prefix sum would be one possible way of computing these offsets, and is an implementation detail. The example below does not compute a prefix sum, it just names a variable `prefix_sum`. The example uses fixed-length splines so the offsets are all just multiples of that length.

Apologies, I know this example has been around for a while, but I think it is not very self-explanatory. It is made up of a bunch of fairly opaque numpy calls to generate data. If instead the example used trivial length splines (e.g. 4 or 5 points), then the data could be explicitly written out and the example would be easier to follow.

**Review by harrism — CHANGES_REQUESTED** [review]

File: python/cuspatial/cuspatial/core/interpolate.py (hunk 3 above)
Comment (isVoid) [reply python/cuspatial/cuspatial/core/interpolate.py]: Fair enough, I updated the example to make it clearer.

File: python/cuspatial/cuspatial/core/interpolate.py [hunk 4]
```diff
@@ -69,33 +69,35 @@ class CubicSpline:
...
-                0, (SPLINE_LENGTH - 1), (SPLINE_LENGTH - 1) * 2 + 1
-            ),) * NUM_SPLINES)
-        ).astype('float32')
-    >>> curve_ids = cudf.Series(np.repeat(
-            np.arange(0, NUM_SPLINES), SPLINE_LENGTH * 2 - 1
-        ), dtype="int32")
-    >>> new_points = curve(new_samples, curve_ids)
+    # The following example only serves as demonstration purposes of
+    # cuspatial.CubicSpline API. cuSpatial does not out perform scipy when
+    # fitting a small number of curves shown in the example below.
```
Comment (harrism) [inline python/cuspatial/cuspatial/core/interpolate.py:79]:
```suggestion
    # The following example only serves to demonstrate the
    # cuspatial.CubicSpline API. cuSpatial does not out perform scipy when
    # fitting a small number of curves as shown.
```

**Review by harrism — CHANGES_REQUESTED** [review]

**Review by thomcom — APPROVED** [review]
lgtm

**Comment by isVoid** [comment]
close #525

**Review by harrism — APPROVED** [review]

**Comment by thomcom** [comment]
@gpucibot merge

**Pull request merged**